sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -598,6 +598,13 @@ fromClause
     : FROM relation (',' relation)* lateralView* pivotClause?
     ;
 
+temporalClause
+    : VERSION AS OF version=INTEGER_VALUE
+    | FOR SYSTEM_VERSION AS OF version=INTEGER_VALUE
```

Comment:
should the version be non-number? e.g. `INTEGER_VALUE | IDENTIFIER`

Reply:
I agree. Shouldn't all of these be strings or even more generic so that the user can pass whatever is a version in a particular storage layer? HBase would use a long, but versions could easily be a UUID or some other type.

Reply:
Yeah, I agree we should allow non-number version here. This is not consistent with `TableCatalog.java:loadTable()` now.

Reply:
Changed to `INTEGER_VALUE | IDENTIFIER` so it can take both number and non-number.

I didn't change it to `String`. Seems to me that `String` type requires single quotation. For example, the syntax needs to be ` SELECT * FROM default.people10m VERSION AS OF '123456789'`. I think we want to have ` SELECT * FROM default.people10m VERSION AS OF 123456789`.

Reply:
You could allow an identifier, string, or integer if you want to avoid requiring quotes.

Reply:
With `version=(INTEGER_VALUE | IDENTIFIER)`, we can have both number and non number, e.g. 
`SELECT * FROM t VERSION AS OF 123456789` 
and 
`SELECT * FROM t VERSION AS OF Snapshot123456789`. 
Do we need to support String syntax such as `SELECT * FROM t VERSION AS OF 'Snapshot123456789'`?

Reply:
I think so. Not everything can be expressed easily as an identifier. For example, if you wanted to specify a version string that is a UUID, the most natural way to do that would be to quote it: `VERSION AS OF '8a22ccf1-9bb6-4a39-99ca-13afad959fe2'`. We could quote with backticks to use an identifier instead, but if we can support either quote style then that reduces user mistakes.

Reply:
Got it. I will add support for String.